### PR TITLE
Create load_data and load_schema APIs [RHELDST-4862]

### PR DESCRIPTION
**Pull request by crungehottman**
The load_schema API allows clients to more easily consume the
schema.json file that is included in the cdn-definitions package.

The load_data API will return a raw dictionary of the chosen data
file. This will allow users to access various types of data in a
uniform way.

In order to encourage a single, consistent method of access to
the cdn-definitions data, the legacy access methods and related
classes (i.e., PathAlias, rhui_alias, and origin_alias) will be
deprecated. A dedicated Python object will no longer need to be
created for each new addition to the schema.

The load_data API also includes an optional "source" argument to
the load_data method. The source parameter allows the user to
specify the source of the cdn_definitions JSON or YAML data file.
The source could be a URL or a local path in a directory tree.

If a source is not specified, the location may be overriden via the
CDN_DEFINITIONS_PATH environment variable, which also may either
be in the form of a URL or a local path.

If neither a source is specified, nor a CDN_DEFINITIONS_PATH
environment variable is set, the data.yaml file included in the
cdn-definitions package will be used as a source. If that file is
unavailable, load_data will attempt to load data from
/usr/share/cdn-definitions/data.yaml. If all of the sources are
invalid, a RuntimeError will be thrown.

### CODE REVIEW

File: src/cdn_definitions/_impl/cdn_data.py
Comment (rohanpm) [inline src/cdn_definitions/_impl/cdn_data.py:4]: About this CdnData class.  I think there's a decision to be made about the future of this library before we can know whether it's right to introduce this class:

For this library I think we have two main approaches how we could expose the data:

1. We could just return exactly what's loaded - a raw dict.
2. We return some kind of model objects which have been created from the underlying data, which could have additional validations, default values etc.

Currently, with rhui_aliases and origin_aliases, this library has gone for approach (2) as it doesn't expose the raw dict.  Instead it returns these PathAlias objects which have 'src', 'dest' attributes and apply some additional checks on top of just the schema.

Even though it's already done that way, I'm not entirely confident that's the best way and there's still a possibility of changing it before moving forward.  But whatever we do, we should be consistent and pick one approach or the other, we really don't want clients using a raw dict for some bits of data and objects for others.

I believe you'd be working on moving some data from cdn-utils into this project soon, so have you thought about what that will look like?  To take an example, `RELEASE_VERSIONS_BY_TYPE`: if we stick with (2), then we shouldn't just add a method/property which returns a raw `release_versions_by_type` dict.  Instead we'd be designing some kind of `ReleaseVersionMapping` class or similar, maybe adding an enum for all the valid types of aus, eus, tus, e4s.  Client may write something like: `CdnData().release_versions.for_type(ReleaseType.EUS)`.

I can't put my finger on exactly why but I'm starting to think that approach (2) might have been a mistake here and we should really just return the dict.  i.e. we migrate to approach (1) before going ahead with adding more cdn-utils data here.

Anyway, back to this CdnData class in this PR.  The impact on this class would be:

- If you want to continue with (2) then this class is basically right, and in the future it'll have to have a lot of additional properties/methods added, and other related classes created.
- Or if you agree with me that we should change direction and go for (1) - then I think this class either shouldn't exist, or it should be a private implementation detail (i.e. it's not imported into cdn_definitions/\_\_init\_\_.py).  The top-level rhui_aliases, origin_aliases functions could still exist temporarily, marked as deprecated, but going forward all clients are suggested to just use load_data which returns a plain dict.

File: src/cdn_definitions/api/__init__.py
Comment (rohanpm) [inline src/cdn_definitions/api/__init__.py:1]: There shouldn't be an "api" module, the API is provided from cdn_definitions module.  No need for an additional "api" layer.

The convention is that everything which is accessible underneath the top-level module, and which doesn't have a name beginning with "_", is public API.

You should stick this code somewhere under _impl, and then in cdn_definitions/\_\_init\_\_.py you should import it - that'll make it public API without exposing anything which should not be exposed.

**Review by rohanpm — CHANGES_REQUESTED** [review]

File: src/cdn_definitions/_impl/__init__.py
Comment (crungehottman) [reply src/cdn_definitions/_impl/__init__.py:21]: I'm probably using an inefficient git workflow, but because I frequently squash and rebase my changes, I find that it's much easier and less confusing to work with a separate branch per commit, rather than with several commits on one branch. So, I would prefer to handle this small change in a separate PR, and rebase this PR's commits on top of that change after it's merged: Ihttps://github.com/release-engineering/cdn-definitions/pull/21

File: src/cdn_definitions/_impl/cdn_data.py
Comment (crungehottman) [reply src/cdn_definitions/_impl/cdn_data.py:4]: >  I believe you'd be working on moving some data from cdn-utils into this project soon, so have you thought about what that will look like? To take an example, RELEASE_VERSIONS_BY_TYPE: if we stick with (2), then we shouldn't just add a method/property which returns a raw release_versions_by_type dict. Instead we'd be designing some kind of ReleaseVersionMapping class or similar, maybe adding an enum for all the valid types of aus, eus, tus, e4s. Client may write something like: CdnData().release_versions.for_type(ReleaseType.EUS).

Yes, I did start fiddling around with creating a schema, set of reference data, and ReleaseVersionMapping class last week. I realized Python objects could potentially make it more difficult to extend the data in cdn-definitions-private. Say, for example, we want to add a new, super-secret "RUS" stream and for whatever reason we are forbidden from releasing any information about the existence of this stream to the public. If the rel-eng wants to configure the additional RUS stream, how would they do so without also adding an RUS attribute in the ReleaseType class? (I suppose they could add a "HIDDEN" attribute or something, but that's besides the point.) Whereas, if the data returned was a dictionary, so long as we allowed for additional properties on the schema's definition for Release Types, we could add an additional release type in our cdn-definitions-private/data.yaml that isn't dependent on a hard-coded set of releases in cdn-definitions. If we wanted to restrict the Release Types, we could instead impose that check by adding a test in cdn-definitions-private. That was a bad example, but hopefully you caught my gist and extrapolate beyond the example.

> I can't put my finger on exactly why but I'm starting to think that approach (2) might have been a mistake here and we should really just return the dict. i.e. we migrate to approach (1) before going ahead with adding more cdn-utils data here.

I agree.  I am having trouble imagining a scenario in which the python objects would offer enough benefit over a dictionary to justify the increased complexity and the potential for additional restraints. Option (1) it is.

File: tox.ini
Comment (crungehottman) [inline tox.ini:13]: Will rebase after https://github.com/release-engineering/cdn-definitions/pull/22 is merged

File: docs/userguide.rst
Comment (rohanpm) [inline docs/userguide.rst:49]:
```suggestion
     if my_path.startswith(alias["src"]):
```

File: docs/userguide.rst
Comment (rohanpm) [inline docs/userguide.rst:79]: Would you mind dropping these from the docs since we are deprecating them so we're only presenting the API that we want people to use going forward?

File: src/cdn_definitions/_impl/__init__.py
Comment (rohanpm) [inline src/cdn_definitions/_impl/__init__.py:27]: I think the method of doing requests needs to be beefed up a little:

- We've got to check status codes on the responses.  If we just try to unconditionally parse them as JSON or YAML then we will throw away error info when the requests fail (see RHELDST-4813 for an example where a tool just tries to parse server response as JSON without checking the status code)
- Could we please apply some modest retries here so we don't immediately fail if the request doesn't work on the first try?  In most cases where we don't add that, we have to come back and add it later.  In this case it would look similar to https://findwork.dev/blog/advanced-usage-python-requests-timeouts-retries-hooks/#retry-on-failure .

File: src/cdn_definitions/_impl/__init__.py
Comment (rohanpm) [inline src/cdn_definitions/_impl/__init__.py:87]: Maybe we can drop the 'global' term, I had mentioned that before since we were going to have a CdnData class and it looked like calling e.g. rhui_alias on an instance of that class, rather than the global one, would still be supported.

Now as we've dropped the CdnData class concept and there's no non-global form of these APIs, we could just delete the word "global" here.

File: tests/test_load.py
Comment (rohanpm) [inline tests/test_load.py:46]: But we aren't setting an env var and it's not always a JSON file.

File: tests/test_load.py
Comment (rohanpm) [inline tests/test_load.py:87]: This looks like it has accidentally been placed outside of any test function, so it's being executed at import time - can we move it to a test function?

**Review by rohanpm — CHANGES_REQUESTED** [review]

File: src/cdn_definitions/_impl/__init__.py
Comment (rohanpm) [inline src/cdn_definitions/_impl/__init__.py:22]: You might consider splitting this into some smaller functions, there's a lot of conditionals here.

**Review by rohanpm — APPROVED** [review]

File: src/cdn_definitions/_impl/__init__.py
Comment (rohanpm) [inline src/cdn_definitions/_impl/__init__.py:12]: Not a big deal, but the function name and doc string aren't actually true since JSON or YAML files can contain kinds of data other than a dict.  e.g. a file containing literally `1` is valid JSON and valid YAML and would cause this function to return an integer.  I'd have named it something like `parsed` since it's parsing the input as JSON or YAML but not converting it to any particular type.

File: src/cdn_definitions/_impl/__init__.py
Comment (rohanpm) [inline src/cdn_definitions/_impl/__init__.py:15]: Doesn't `data.json()` already parse the content, so you're trying to parse it twice here?

If I'm right then you'd probably want to also look into how it is that the tests could pass while this bug exists.

File: tests/test_load.py
Comment (rohanpm) [inline tests/test_load.py:62]: Regarding the earlier comment about wrong use of `json.loads(data.json(), ...)`

I see the issue here in the test.  The `json` here is meant to be *an object which will be serialized as JSON*, not a JSON string.  So this should have been:

```suggestion
        m.get("https://test.com/data.json", json={"green": "bean"})
```

requests_mock accepts the string since strings can also be serialized to JSON, so you've set up a "JSON within JSON" response here, which makes it so that the double-decoding in the code under test happens to work.

**Review by rohanpm — CHANGES_REQUESTED** [review]
Could you comment about the release of this - do you want it to be released immediately once it's approved?

File: tests/test_load.py
Comment (crungehottman) [reply tests/test_load.py:62]: Whoops! That explains the error that provoked me to use json.loads(data.json()) in the first place. Thanks!

File: src/cdn_definitions/_impl/__init__.py
Comment (crungehottman) [reply src/cdn_definitions/_impl/__init__.py:12]: Ah, that's true. Changed the function name and doc string.

**Comment by crungehottman** [comment]
> Could you comment about the release of this - do you want it to be released immediately once it's approved?

Well, progress on https://github.com/release-engineering/exodus-lambda/pull/116 and merging of https://gitlab.cee.redhat.com/cdn/cdn-definitions-private/-/merge_requests/6 will be blocked until a new release of cdn-definitions lands in PyPI. I still need to add the schemas (etc) though, and I don't want to burden you (? I don't think I can release this) with several releases, so I will add the schemas first--I think the release can wait until the schemas have been merged.

File: src/cdn_definitions/_impl/__init__.py
Comment (rohanpm) [inline src/cdn_definitions/_impl/__init__.py:92]:
```suggestion
    Loads the `schema.json` file provided with the cdn-definitions package into a dictionary.
```

**Review by rohanpm — DISMISSED** [review]

File: src/cdn_definitions/_impl/__init__.py
Comment (crungehottman) [reply src/cdn_definitions/_impl/__init__.py]: This comment made me realize that I forgot to rephrase this line that suggests that this method will always return a dict, as you pointed out, here: 

>  Not a big deal, but the function name and doc string aren't actually true since JSON or YAML files can contain kinds of data other than a dict. e.g. a file containing literally 1 is valid JSON and valid YAML and would cause this function to return an integer. I'd have named it something like parsed since it's parsing the input as JSON or YAML but not converting it to any particular type.


 I changed `load_data`'s and `load_schema`'s doc strings to more generally state that they return the data coerced into a python object instead of a dict.

**Comment by crungehottman** [comment]
Sorry, @nathanegillett @lipoja - I neglected to CC you here earlier. Again, I cannot add either of you as a reviewer.

File: src/cdn_definitions/_impl/__init__.py
Comment (nathanegillett) [inline src/cdn_definitions/_impl/__init__.py:35]: Isn't this equivalent to writing the following?
```
response.raise_for_status()
return response
```

**Review by nathanegillett — DISMISSED** [review]

**Review by rohanpm — APPROVED** [review]
Any reasons not to submit this yet?  (Doesn't have to be *released* yet.)

**Review by nathanegillett — APPROVED** [review]

**Pull request merged**
